docs/ADR-8-Use-RIO-in-cardano‐cli.md

```diff
@@ -0,0 +1,129 @@
+
+![image](https://github.com/user-attachments/assets/6a50e8f2-9140-4b7e-afa5-ff778d696742)
+ 
+- The ExceptT IO is a known anti-pattern for these reasons and others as per:
+  - https://github.com/haskell-effectful/effectful/blob/master/transformers.md
+  - https://tech.fpcomplete.com/blog/2016/11/exceptions-best-practices-haskell/
+
+
+# Proposed Solution
+
+Use the RIO monad which is based on the [ReaderT design pattern](https://tech.fpcomplete.com/blog/2017/06/readert-design-pattern/).
+
+### Pros 
+I opt for the `RIO env a` monad instead of `ReaderT e m a` because: 
+- We get additional [logging functionality](https://hackage.haskell.org/package/rio-0.1.22.0/docs/RIO.html#g:5)
+- Explicit environment dependencies e.g `logError :: HasLogFunc env => Text -> RIO env ()`
+- Better composability i.e no more errors that wrap errors (see above).
+
+### Cons
+- `RIO` is hardcoded to IO so we cannot add additional transformer layers e.g `RIO Env (StateT Int IO) a`
+- Implicit error flow. Errors are thrown via `IO` exceptions.
+
+### Exception handling
+
+#### Exception type
+```haskell
+data CustomCliException where
+  CustomCliException
+    :: (Show error, Typeable error, Error error, HasCallStack)
+    => error -> CustomCliException
+
+deriving instance Show CustomCliException
+
+instance Exception CustomCliException where
+  displayException (CustomCliException e) =
+    unlines
+      [ show (prettyError e)
+      , prettyCallStack callStack
+      ]
+```
+
+#### Top level handler 
+```haskell
+executeRio :: RIO () () -> IO ()
+executeRio r = do
+  runRIO () r
+    `catch` ( \e ->
+                case fromException e of
+                  Just custom@(CustomCliException{}) -> putStrLn $ displayException custom
+                  Nothing -> putStrLn $ displayException e
+            )
+```
+
+### Exception Handling Mechanism: Pros & Cons
+
+#### Pros
+
+1. **Unified Exception Type**
+```haskell
+data CustomCliException where
+  CustomCliException :: (Show e, Typeable e, Error e, HasCallStack) => e -> CustomCliException
+
+deriving instance Show CustomCliException
+
+instance Exception CustomCliException where
+  displayException (CustomCliException e) =
+    unlines
+      [ show (prettyError e)
+      , prettyCallStack callStack
+      ]
+```
+- Simplifies Top-Level Handling: All errors are caught as `CustomCliException`.
+- Consistent Reporting: Ensures all errors are formatted uniformly via `prettyError`.
+
+2. **CallStack Inclusion**
+
+- Embeds CallStack to trace error origins, aiding debugging.
+
+3. **Polymorphic Error Support**
+
+- Flexibility: Wraps any error type so long at the required instances exist.
+
+#### Cons
+1. **Type Erasure**
+
+- Loss of Specificity: Existential quantification erases concrete error types, preventing pattern-matching on specific errors.
+
+# Decision
+- Error handling: All errors will be converted to exceptions that will be caught by a single exception handler at the top level.
```

Comment:
Converted how? This is not explained in the examples.

Reply:
https://github.com/input-output-hk/cardano-node-wiki/pull/60/files#diff-c83af4b681074eae4a9a579fb6a393539599918fd6e7c9e069d8f2422590b48aR100

Reply:
Linking to a whole PR from an ADR is not ideal:
1. That PR is too big to illustrate how it should be used. Now a reader has to go through that PR, sift through non-relevant parts, read it, and figure out what's the idea here.
2. A PR until it's merged is a living thing, can still change.
3. There's a potential bitrot possibility in case yet another github organisation migration happens, github.com dissapearance etc.

Could you provide a self-contained example what do you mean here? 

Additionally that PR does not show any errors' conversion.

Do you mean using exceptions as a replacement for `Except e m a` here?